source/simulator/core.cc

```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+               */
+              do
+                {
+                  current_linearization_point = backup_linearization_point;
+
+                  LinearAlgebra::BlockVector search_direction = solution;
+
+                  search_direction *= lambda;
+
+
+                  current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
+                  current_linearization_point.block(introspection.block_indices.velocities) += search_direction.block(introspection.block_indices.velocities);
+
+                  assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
+                  rebuild_stokes_matrix = true;
+
+                  assemble_stokes_system();
+
+                  test_velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  test_pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  test_residual = std::sqrt(test_velocity_residual * test_velocity_residual
+                                            + test_pressure_residual * test_pressure_residual);
+
+                  if (test_residual < (1.0 - alpha * lambda) * residual
+                      ||
+                      line_search_iteration >= parameters.max_newton_line_search_iterations
+                      ||
+                      use_picard)
+                    {
+                      pcout << "      Total relative residual after nonlinear iteration " << nonlinear_iteration+1
+                            << ": " << residual/initial_residual << ", norm of the rhs: " << test_residual
+                            << ", newton_derivative_scaling_factor: " << newton_handler->get_newton_derivative_scaling_factor() << std::endl;
+                      break;
+                    }
+                  else
+                    {
+
+                      pcout << "   Line search iteration " << line_search_iteration << ", with norm of the rhs "
+                            << test_residual << " and going to " << (1.0 - alpha * lambda) * residual
+                            << ", relative residual: " << residual/initial_residual << std::endl;
+
+                      /**
+                       * The line search step was not sufficient to decrease the residual
+                       * enough, so we take a smaller step to see if it improves the residual.
+                       */
+                      lambda *= (2.0/3.0);// TODO: make a parameter out of this.
+                    }
+
+                  line_search_iteration++;
+                  Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
+                         ExcMessage ("This tests the while condition. This condition should "
+                                     "actually never be false, because the break statement "
+                                     "above should have caught it."));
```

Comment:
This is not a useful message to users. And the condition may happen -- you just haven't found a sufficiently small update that satisfies the stopping condition. This can happen for poorly conditioned problems and/or insufficiently solved linear systems.

I think I would just say something like "The line search as part of the Newton method did not find a suitable update to the current solution within the allowed number of steps. Try allowing more steps, solving the linear system more accurately, or reducing the degree of ill-posedness of the original problem."